File: Tarea1/Stage4/Broker.java
import java.util.ArrayList;

public class Broker {
   public Broker() {
      topics = new ArrayList<Topic>();
   }
   public Topic createTopic (String topicName){
      //....................................................................
      Topic topic =new Topic(topicName);  
      topics.add(topic);
      //....................................................................
      return topic;
   }
   public boolean subscribe (Subscriber sub){
      //....................................................................
      Topic topic=createTopic(sub.topicName);
      String topicName=sub.topicName;
      //....................................................................
      if ((topic=findTopic(topicName))!=null) {
         topic.subscribe(sub);
         return true;
      } else 
         return false; // topic does not exist.
   }
   private Topic findTopic (String topicName) {
      //....................................................................
      for (Topic p : topics){
         if(p.hasThisName(topicName)){
            return p;
         }
      }
      //....................................................................
      return null;  // topic not found
   }
   private ArrayList<Topic> topics;
}

File: Tarea3/Stage3/broker.cpp
#include "broker.h"
#include "topic.h"
#include "subscriber.h"

Broker::Broker() {
    // Constructor vacío, inicialización de topics ya está en el campo
}

Topic* Broker::createTopic(const QString& topicName) {
    Topic* topic = findTopic(topicName);
    if (!topic) {
        topic = new Topic(topicName);
        topics.append(topic);
    }
    return topic;
}

bool Broker::subscribe(Subscriber* sub) {
    QString topicName = sub->getTopicName();
    Topic* topic = findTopic(topicName);
    if (topic) {
        topic->subscribe(sub);
        return true;
    } else {
        return false; // tópico no existe
    }
}

Topic* Broker::findTopic(const QString& topicName) {
    for (Topic* topic : topics) {
        if (topic->hasThisName(topicName)) {
            return topic;
        }
    }
    return nullptr;
}
